### PR TITLE
Add gha_update_pr_comment foundation (POST/PATCH in github_actions_api)

### PR DESCRIPTION
**Pull request by amd-hsivasun**
## Motivation
ISSUE ID : https://github.com/ROCm/rocm-systems/issues/6597

When a TheRock submodule bump merges, upstream PRs in `rocm-systems` and `rocm-libraries` have no signal that their changes are now part of TheRock. Post-merge breadcrumb comments on those PRs need a shared way to create or update idempotent PR comments from CI.

This PR lands that foundation only: `gha_update_pr_comment()` in `github_actions_api.py`. Wiring into `bump_submodules.yml` (breadcrumbs) is a follow-on PR. The helper is also intended as a reusable alternative to ad-hoc upsert logic elsewhere (e.g. `therock_pr_bot/policy_check.py`).

## Technical Details
This PR lands the foundation only: extend github_actions_api with write support and add gha_update_pr_comment(). 
Follow-on PRs (not in this PR):
- Bump breadcrumbs: `post_bump_breadcrumbs.py` + step in `bump_submodules.yml`
- Manifest-diff report discoverability: step summary link (does not use this helper)

- Extend  gha_send_request with keyword-only method and body for POST/PATCH (REST and gh api paths).
- Add gha_update_pr_comment() to create or update a sticky PR comment identified by an HTML marker (paginated GET → PATCH or POST).
- Add unit tests for POST/PATCH plumbing, upsert paths, pagination, and GET empty-body backward compatibility.
- Document auth expectations in the helper docstring: callers set GITHUB_TOKEN on the step (GITHUB_TOKEN: ${{ github.token }} for in-repo; App token override for cross-repo). Posting comments requires pull-requests: write in job permissions.

## Test Plan

- pytest build_tools/github_actions/tests/github_actions_api_test.py -v

## Test Result

- 47 passed locally in build_tools/github_actions/tests/github_actions_api_test.py

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: build_tools/github_actions/github_actions_api.py [hunk 1]
```diff
@@ -531,7 +605,138 @@ def gha_send_request(url: str, timeout_seconds: int = 300) -> object:
...
+    """Create or update a sticky PR comment identified by an HTML marker.
+
+    Paginates through existing issue comments on the pull request. If a comment
+    whose body contains ``marker`` is found, that comment is updated in place.
+    Otherwise a new comment is posted.
+
+    Args:
+        pr_number: Pull request number.
+        marker: Unique HTML comment marker embedded in the comment body
+            (e.g. ``<!-- therock-report-manifest-diff -->``).
```
Comment (ScottTodd) [inline build_tools/github_actions/github_actions_api.py]: We could also check the username (`user.login` ?) that authored the comment so the bot doesn't risk editing comments from human user accounts too (e.g. if they quote-reply the bot comment).

File: build_tools/github_actions/github_actions_api.py [hunk 2]
```diff
@@ -531,7 +605,138 @@ def gha_send_request(url: str, timeout_seconds: int = 300) -> object:
...
+    ``pull-requests: write`` (or ``issues: write``) in job ``permissions``.
+    For cross-repo comments, pass ``github_api=GitHubAPI(github_token=...)``
+    with an App installation token that has access to ``github_repository``.
+
+    See: https://docs.github.com/en/rest/issues/comments
+    """
+    if marker not in body:
+        _log(
+            f"Warning: marker {marker!r} not found in comment body for PR #{pr_number}"
+        )
```
Comment (ScottTodd) [inline build_tools/github_actions/github_actions_api.py:664]: Can you add `github_repository` to this log line? Just a PR number isn't enough to unique identify a URL

How about logs like this?
```
Warning: marker
  <!-- therock-report-manifest-diff -->
not found in body for PR comment on https://github.com/ROCm/TheRock/pull/6749
body text:
  <the body text>
```

File: build_tools/github_actions/github_actions_api.py [hunk 3]
```diff
@@ -531,7 +605,138 @@ def gha_send_request(url: str, timeout_seconds: int = 300) -> object:
...
+    while page <= MAX_PAGES:
+        page_url = f"{comments_url}?per_page={PER_PAGE}&page={page}"
+        comments = api.send_request(page_url)
+        if not isinstance(comments, list):
+            break
+
+        for comment in comments:
+            if marker in comment.get("body", ""):
+                existing_comment_id = comment["id"]
+                break
```
Comment (ScottTodd) [inline build_tools/github_actions/github_actions_api.py]: Per my other comment, could also check the comment authorship information here.

If a `comment_author` argument is provided, skip comments that have a different author

File: build_tools/github_actions/tests/github_actions_api_test.py [hunk 4]
```diff
@@ -473,6 +508,297 @@ def test_rest_api_invalid_json_raises_github_api_error(self):
...
+        self.assertEqual(result, {"id": 7})
+        cmd = subprocess_run.call_args[0][0]
+        self.assertEqual(cmd[0:4], ["/usr/bin/gh", "api", "--method", "POST"])
+        self.assertEqual(cmd[4], "/repos/test/test/issues/1/comments")
+        self.assertIn("--input", cmd)
+        self.assertEqual(subprocess_run.call_args[1]["input"], '{"body": "hello"}')
+
+    def test_rest_api_get_empty_body_raises_github_api_error(self):
+        """REST API GET with empty body should raise GitHubAPIError (unchanged behavior)."""
+        import json
```
Comment (ScottTodd) [inline build_tools/github_actions/tests/github_actions_api_test.py]: This import is already there at the top of the file.

Follow the style guide: https://github.com/ROCm/TheRock/blob/main/docs/development/style_guides/python_style_guide.md#import-organization

File: build_tools/github_actions/tests/github_actions_api_test.py [hunk 5]
```diff
@@ -473,6 +508,297 @@ def test_rest_api_invalid_json_raises_github_api_error(self):
...
+            )
+
+        self.assertEqual(result, {})
+
+
+class GhaUpdatePrCommentTest(unittest.TestCase):
+    """Tests for gha_update_pr_comment."""
+
+    def test_posts_new_comment_when_marker_not_found(self):
+        marker = "<!-- therock-report-manifest-diff -->"
```
Comment (ScottTodd) [inline build_tools/github_actions/tests/github_actions_api_test.py]: nit: could use text like `example-marker` instead of `therock-report-manifest-diff` here, to ensure that the code is generic and not hardcoded to only the currently expected manifest diff usage

(throughout the new code)

File: build_tools/github_actions/tests/github_actions_api_test.py [hunk 6]
```diff
@@ -473,6 +508,297 @@ def test_rest_api_invalid_json_raises_github_api_error(self):
...
+    def test_raises_when_comment_response_is_not_dict(self):
+        marker = "<!-- therock-report-manifest-diff -->"
+        body = f"{marker}\nbody"
+
+        with mock.patch(
+            "github_actions_api._default_github_api.send_request",
+            side_effect=[[], "not-a-dict"],
+        ):
+            with self.assertRaisesRegex(GitHubAPIError, "Expected comment object"):
+                gha_update_pr_comment(pr_number=1, marker=marker, body=body)
```
Comment (ScottTodd) [inline build_tools/github_actions/tests/github_actions_api_test.py]: I don't think this is a useful test case. All it checks is that _if_ the GitHub API returns some unexpected data we raise an error.

File: build_tools/github_actions/tests/github_actions_api_test.py [hunk 7]
```diff
@@ -473,6 +508,297 @@ def test_rest_api_invalid_json_raises_github_api_error(self):
...
+                pr_number=9,
+                marker=marker,
+                body=body,
+                github_repository="ROCm/rocm-systems",
+                github_api=custom_api,
+            )
+
+        self.assertEqual(result["id"], 55)
+        self.assertEqual(custom_api.send_request.call_count, 2)
+        singleton_send_request.assert_not_called()
```
Comment (ScottTodd) [inline build_tools/github_actions/tests/github_actions_api_test.py]: This is testing implementation details. I think the `test_explicit_github_token_skips_auto_detection` test case above should be sufficient here.

File: build_tools/github_actions/tests/github_actions_api_test.py [hunk 8]
```diff
@@ -473,6 +508,297 @@ def test_rest_api_invalid_json_raises_github_api_error(self):
...
+
+        self.assertEqual(result, [])
+
+    def test_raises_when_response_is_not_list(self):
+        with mock.patch(
+            "github_actions_api._default_github_api.send_request",
+            return_value={"message": "not found"},
+        ):
+            with self.assertRaisesRegex(GitHubAPIError, "Expected a list"):
+                gha_query_prs_for_commit("ROCm/TheRock", "c" * 40)
```
Comment (ScottTodd) [inline build_tools/github_actions/tests/github_actions_api_test.py]: Similar comment to the `test_raises_when_comment_response_is_not_dict` test case. Just testing return types for APIs outside our control is not really useful here.

File: build_tools/github_actions/tests/github_actions_api_test.py [hunk 9]
```diff
@@ -473,6 +508,297 @@ def test_rest_api_invalid_json_raises_github_api_error(self):
...
+        ) as singleton_send_request:
+            result = gha_query_prs_for_commit(
+                "ROCm/rocm-libraries", sha, github_api=custom_api
+            )
+
+        self.assertEqual(result, prs)
+        custom_api.send_request.assert_called_once_with(
+            f"https://api.github.com/repos/ROCm/rocm-libraries/commits/{sha}/pulls"
+        )
+        singleton_send_request.assert_not_called()
```
Comment (ScottTodd) [inline build_tools/github_actions/tests/github_actions_api_test.py]: Similar comment to the other `test_uses_provided_github_api_instance_instead_of_singleton` above. This is a lot of boilerplate code for minimal actual test coverage when other test cases already exercise the meaningful code paths.

**Comment by amd-hsivasun** [comment]
Thanks for Review @ScottTodd, addressed your points.

**Review by ScottTodd — APPROVED** [review]
